File: editor-lsp/src/main/java/io/github/rosemoe/sora/lsp/editor/LspEditor.java
/*
 *    sora-editor - the awesome code editor for Android
 *    https://github.com/Rosemoe/sora-editor
 *    Copyright (C) 2020-2023  Rosemoe
 *
 *     This library is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU Lesser General Public
 *     License as published by the Free Software Foundation; either
 *     version 2.1 of the License, or (at your option) any later version.
 *
 *     This library is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public
 *     License along with this library; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 *     USA
 *
 *     Please contact Rosemoe by email 2073412493@qq.com if you need
 *     additional information or have any questions
 */
package io.github.rosemoe.sora.lsp.editor;

import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;

import org.eclipse.lsp4j.Diagnostic;
import org.eclipse.lsp4j.FormattingOptions;
import org.eclipse.lsp4j.SignatureHelp;
import org.eclipse.lsp4j.TextDocumentSyncKind;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.TimeoutException;
import java.util.function.Supplier;

import io.github.rosemoe.sora.annotations.Experimental;
import io.github.rosemoe.sora.event.ContentChangeEvent;
import io.github.rosemoe.sora.lang.Language;
import io.github.rosemoe.sora.lsp.client.languageserver.requestmanager.RequestManager;
import io.github.rosemoe.sora.lsp.client.languageserver.serverdefinition.LanguageServerDefinition;
import io.github.rosemoe.sora.lsp.client.languageserver.wrapper.LanguageServerWrapper;
import io.github.rosemoe.sora.lsp.editor.event.LspEditorContentChangeEventReceiver;
import io.github.rosemoe.sora.lsp2.editor.signature.SignatureHelpWindow;
import io.github.rosemoe.sora.lsp.operations.Provider;
import io.github.rosemoe.sora.lsp.operations.completion.CompletionProvider;
import io.github.rosemoe.sora.lsp.operations.diagnostics.PublishDiagnosticsProvider;
import io.github.rosemoe.sora.lsp.operations.diagnostics.QueryDocumentDiagnosticsProvider;
import io.github.rosemoe.sora.lsp.operations.document.ApplyEditsProvider;
import io.github.rosemoe.sora.lsp.operations.document.DocumentChangeProvider;
import io.github.rosemoe.sora.lsp.operations.document.DocumentCloseProvider;
import io.github.rosemoe.sora.lsp.operations.document.DocumentOpenProvider;
import io.github.rosemoe.sora.lsp.operations.document.DocumentSaveProvider;
import io.github.rosemoe.sora.lsp.operations.format.FullFormattingProvider;
import io.github.rosemoe.sora.lsp.operations.format.RangeFormattingProvider;
import io.github.rosemoe.sora.lsp.operations.signature.SignatureHelpProvider;
import io.github.rosemoe.sora.lsp.requests.Timeout;
import io.github.rosemoe.sora.lsp.requests.Timeouts;
import io.github.rosemoe.sora.widget.CodeEditor;

@Experimental
public class LspEditor {

    private final String projectPath;

    private final LanguageServerDefinition serverDefinition;

    private final String currentFileUri;
    private final LspEditorManager lspEditorManager;

    private WeakReference<CodeEditor> currentEditor;

    private WeakReference<SignatureHelpWindow> signatureHelpWindowWeakReference;

    private Language wrapperLanguage;

    private LspLanguage currentLanguage;

    private LspProviderManager providerManager;

    private LanguageServerWrapper languageServerWrapper;

    private Boolean isClose = false;

    private Runnable unsubscribeFunction = null;

    private TextDocumentSyncKind textDocumentSyncKind;

    private List<String> completionTriggers = Collections.emptyList();

    private List<String> signatureHelpTriggers = Collections.emptyList();

    private List<String> signatureHelpRetriggers = Collections.emptyList();

    private LspEditorContentChangeEventReceiver editorContentChangeEventReceiver;


    public LspEditor(String currentProjectPath, String currentFileUri, LanguageServerDefinition serverDefinition, LspEditorManager lspEditorManager) {
        this.currentEditor = new WeakReference<>(null);
        this.providerManager = new LspProviderManager(this);
        this.currentLanguage = new LspLanguage(this);

        this.lspEditorManager = lspEditorManager;
        this.currentFileUri = currentFileUri;
        this.projectPath = currentProjectPath;
        this.serverDefinition = serverDefinition;

        this.editorContentChangeEventReceiver = new LspEditorContentChangeEventReceiver(this);

    }

    public String getCurrentFileUri() {
        return currentFileUri;
    }

    public String getProjectPath() {
        return projectPath;
    }

    public void setEditor(CodeEditor currentEditor) {
        this.currentEditor = new WeakReference<>(currentEditor);

        if (unsubscribeFunction != null) {
            unsubscribeFunction.run();
        }
        currentEditor.setEditorLanguage(currentLanguage);
        signatureHelpWindowWeakReference = new WeakReference<>(new SignatureHelpWindow(currentEditor));

        var subscriptionReceipt = currentEditor.subscribeEvent(ContentChangeEvent.class, editorContentChangeEventReceiver);

        unsubscribeFunction = subscriptionReceipt::unsubscribe;

    }

    @Nullable
    public CodeEditor getEditor() {
        return currentEditor.get();
    }

    public LspLanguage getLanguage() {
        return currentLanguage;
    }

    @Nullable
    public <T extends Language> T getWrapperLanguage() {
        return (T) wrapperLanguage;
    }

    /**
     * Set the wrapper language, as the language server may not provide all the features, such as highlighting, you can implement the features yourself and integrate with the language server
     */
    public void setWrapperLanguage(Language wrapperLanguage) {
        this.wrapperLanguage = wrapperLanguage;
        currentLanguage.setWrapperLanguage(wrapperLanguage);
        var editor = currentEditor.get();
        if (editor != null) {
            setEditor(editor);
        }
    }

    /**
     * Return the provider manager to manage this editor
     */
    public LspProviderManager getProviderManager() {
        return providerManager;
    }

    /**
     * @deprecated recommended to use {@link LspProviderManager#addProvider(Supplier)}
     */
    @Deprecated
    public void installFeature(Supplier<Provider<?, ?>> featureSupplier) {
        providerManager.addProvider(featureSupplier);
    }

    /**
     * @deprecated recommended to use {@link LspProviderManager#addProviders(Supplier[])}
     */
    @Deprecated
    @SafeVarargs
    public final void installFeatures(Supplier<Provider<?, ?>>... featureSupplier) {
        providerManager.addProviders(featureSupplier);
    }

    /**
     * Remove a feature for the given class
     *
     * @deprecated The internal implementation has been migrated to {@link LspProviderManager} and  recommended to use {@link LspProviderManager#removeProvider(Class)}
     */
    @Deprecated
    public void uninstallFeature(Class<?> featureClass) {
        providerManager.removeProvider(featureClass);
    }

    /**
     * Return a feature instance for the given class
     *
     * @deprecated The internal implementation has been migrated to {@link LspProviderManager} and recommended to use {@link LspProviderManager#useProvider(Class)}
     */
    @Nullable
    @Deprecated
    public <T extends Provider> T useFeature(Class<T> featureClass) {
        return providerManager.useProvider(featureClass);
    }

    /**
     * Return feature instances based on Optional wrappers.
     *
     * @deprecated The internal implementation has been migrated to {@link LspProviderManager} and recommended to use {@link LspProviderManager#safeUseProvider(Class)}
     */
    @Deprecated
    public <T extends Provider> Optional<T> safeUseFeature(Class<T> featureClass) {
        return providerManager.safeUseProvider(featureClass);
    }


    public void installFeatures() {

        //features
        providerManager.addProviders(RangeFormattingProvider::new, DocumentOpenProvider::new, DocumentSaveProvider::new, DocumentChangeProvider::new, DocumentCloseProvider::new, PublishDiagnosticsProvider::new, CompletionProvider::new, FullFormattingProvider::new, ApplyEditsProvider::new,
                QueryDocumentDiagnosticsProvider::new, SignatureHelpProvider::new);

        //options

        // formatting
        var formattingOptions = new FormattingOptions();
        formattingOptions.setTabSize(4);
        formattingOptions.setInsertSpaces(true);
        providerManager.addOption(formattingOptions);

    }


    /**
     * @deprecated recommended to use {@link LspProviderManager#getOption(Class)}
     */
    @Deprecated
    @Nullable
    public <T> T getOption(Class<T> optionClass) {
        return providerManager.getOption(optionClass);
    }


    private void setupLanguageServerWrapper() {
        var languageServerWrapper = LanguageServerWrapper.forProject(projectPath);
        languageServerWrapper = languageServerWrapper != null ? languageServerWrapper : new LanguageServerWrapper(serverDefinition, projectPath);
        languageServerWrapper.serverDefinition = serverDefinition;
        this.languageServerWrapper = languageServerWrapper;
    }

    /**
     * Connect to the language server to provide the capabilities, this will cause threads blocking. Note: An error will be thrown if the language server is not connected after some time.
     *
     * @see io.github.rosemoe.sora.lsp.requests.Timeouts
     * @see io.github.rosemoe.sora.lsp.requests.Timeout
     */
    @WorkerThread
    public void connect() throws TimeoutException {

        setupLanguageServerWrapper();

        languageServerWrapper.start();
        //wait for language server start
        var server = languageServerWrapper.getServer();
        if (server == null) {
            throw new TimeoutException("Unable to connect language server");
        }
        languageServerWrapper.connect(this);
    }

    /**
     * Try to connect to the language server repeatedly, this will cause threads blocking. Note: An error will be thrown if the language server is not connected after some time.
     *
     * @see io.github.rosemoe.sora.lsp.requests.Timeouts
     * @see io.github.rosemoe.sora.lsp.requests.Timeout
     */
    @WorkerThread
    public void connectWithTimeout() throws InterruptedException, TimeoutException {

        setupLanguageServerWrapper();

        var start = System.currentTimeMillis();
        var retryTime = Timeout.getTimeout(Timeouts.INIT);
        long maxRetryTime = start + retryTime;

        while (start < maxRetryTime) {
            try {
                connect();
                break;
            } catch (Exception exception) {
                //exception.printStackTrace();
            }
            start = System.currentTimeMillis();

            Thread.sleep(retryTime / 10);

        }

        if (start > maxRetryTime) {
            throw new TimeoutException("Unable to connect language server");
        } else {
            connect();
        }

    }


    public String getEditorContent() {
        return currentEditor.get().getText().toString();
    }


    @Nullable
    public List<Diagnostic> getDiagnostics() {
        return lspEditorManager.diagnosticsContainer.getDiagnostics(currentFileUri);
    }

    public void onDiagnosticsUpdate() {
        publishDiagnostics(getDiagnostics());
    }

    private void publishDiagnostics(List<Diagnostic> diagnostics) {
        getProviderManager().safeUseProvider(PublishDiagnosticsProvider.class)
                .ifPresent(publishDiagnosticsFeature -> publishDiagnosticsFeature.execute(diagnostics));
    }


    /**
     * Notify the language server to open the document
     */
    public void open() {
        getProviderManager().safeUseProvider(DocumentOpenProvider.class).ifPresent(documentOpenFeature -> documentOpenFeature.execute(null));
    }


    /**
     * Notify language servers to save document
     */
    public void save() {
        getProviderManager().safeUseProvider(DocumentSaveProvider.class).ifPresent(documentSaveFeature -> documentSaveFeature.execute(null));
    }


    /**
     * Get a request manager that can manipulate the language server
     */
    @Nullable
    public RequestManager getRequestManager() {
        LanguageServerWrapper serverWrapper = LanguageServerWrapper.forProject(projectPath);
        return serverWrapper != null ? serverWrapper.getRequestManager() : null;
    }

    public Optional<RequestManager> getRequestManagerOfOptional() {
        return Optional.ofNullable(getRequestManager());
    }

    /**
     * disconnect to the language server
     */
    public void disconnect() {
        if (languageServerWrapper != null) {
            try {
                var feature = getProviderManager().useProvider(DocumentCloseProvider.class);
                if (feature != null) feature.execute(null).get();

                ForkJoinPool.commonPool().execute(() -> languageServerWrapper.disconnect(this));

            } catch (ExecutionException | InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public void showSignatureHelp(SignatureHelp signatureHelp) {
        var signatureHelpWindow = signatureHelpWindowWeakReference.get();
        if (signatureHelpWindow == null) {
            return;
        }
        var editor = currentEditor.get();
        if (editor == null) {
            return;
        }
        if (signatureHelp == null) {
            editor.post(signatureHelpWindow::dismiss);
            return;
        }
        editor.post(() -> signatureHelpWindow.show(signatureHelp));
    }


    private void dispose() {
        if (languageServerWrapper != null) {
            languageServerWrapper.unregister();
        }
        providerManager.dispose();

        currentEditor.clear();
        completionTriggers.clear();

        currentLanguage.destroy();

        currentLanguage = null;
        providerManager = null;
        completionTriggers = null;

        if (unsubscribeFunction != null) {
            unsubscribeFunction.run();
            unsubscribeFunction = null;
        }

        editorContentChangeEventReceiver = null;
    }

    /**
     * Close the editor, it will dispose the editor's resources and disconnect to the language server
     */
    public void close() {

        if (isClose) {
            return;
        }

        isClose = true;

        disconnect();
        dispose();
    }


    public TextDocumentSyncKind getSyncOptions() {
        return textDocumentSyncKind == null ? TextDocumentSyncKind.Full : textDocumentSyncKind;
    }

    public void setSyncOptions(TextDocumentSyncKind textDocumentSyncKind) {
        this.textDocumentSyncKind = textDocumentSyncKind;
    }

    public void setCompletionTriggers(List<String> completionTriggers) {
        this.completionTriggers = new ArrayList<>(completionTriggers);
    }


    public String getFileExt() {
        return serverDefinition.ext;
    }

    public List<String> getCompletionTriggers() {
        return this.completionTriggers;
    }

    public void setSignatureHelpTriggers(List<String> signatureHelpTriggers) {
        this.signatureHelpTriggers = new ArrayList<>(signatureHelpTriggers);
    }

    public List<String> getSignatureHelpTriggers() {
        return this.signatureHelpTriggers;
    }

    public void setSignatureHelpRetriggers(List<String> signatureHelpRetriggers) {
        if (signatureHelpRetriggers.size() < 1 && signatureHelpTriggers.size() > 0 && signatureHelpTriggers.contains("(")) {
            this.signatureHelpRetriggers = List.of(")");
            return;
        }
        this.signatureHelpRetriggers = signatureHelpRetriggers;
    }

    public List<String> getSignatureHelpRetriggers() {
        return this.signatureHelpRetriggers;
    }

    public boolean hitRetrigger(CharSequence eventText) {
        for (var trigger : signatureHelpRetriggers) {
            if (trigger.contains(eventText)) {
                return true;
            }
        }
        return false;
    }

    public boolean hitTrigger(CharSequence eventText) {
        for (var trigger : signatureHelpTriggers) {
            if (trigger.contains(eventText)) {
                return true;
            }
        }
        return false;
    }

    public boolean isShowSignatureHelp() {
        var signatureHelpWindow = signatureHelpWindowWeakReference.get();
        if (signatureHelpWindow == null) {
            return false;
        }
        return signatureHelpWindow.isShowing();
    }


}

File: editor-lsp/src/main/java/io/github/rosemoe/sora/lsp/editor/LspEditor.kt
/*******************************************************************************
 *    sora-editor - the awesome code editor for Android
 *    https://github.com/Rosemoe/sora-editor
 *    Copyright (C) 2020-2023  Rosemoe
 *
 *     This library is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU Lesser General Public
 *     License as published by the Free Software Foundation; either
 *     version 2.1 of the License, or (at your option) any later version.
 *
 *     This library is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public
 *     License along with this library; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 *     USA
 *
 *     Please contact Rosemoe by email 2073412493@qq.com if you need
 *     additional information or have any questions
 ******************************************************************************/

package io.github.rosemoe.sora.lsp.editor

import androidx.annotation.WorkerThread
import io.github.rosemoe.sora.annotations.Experimental
import io.github.rosemoe.sora.lang.Language
import io.github.rosemoe.sora.lsp.client.languageserver.requestmanager.RequestManager
import io.github.rosemoe.sora.lsp.client.languageserver.wrapper.LanguageServerWrapper
import io.github.rosemoe.sora.lsp.editor.format.LspFormatter
import io.github.rosemoe.sora.lsp.events.EventType
import io.github.rosemoe.sora.lsp.events.diagnostics.publishDiagnostics
import io.github.rosemoe.sora.lsp.events.document.documentClose
import io.github.rosemoe.sora.lsp.events.document.documentOpen
import io.github.rosemoe.sora.lsp.events.document.documentSave
import io.github.rosemoe.sora.lsp.events.navigation.definition
import io.github.rosemoe.sora.lsp.events.navigation.references
import io.github.rosemoe.sora.lsp.events.progress.cancelProgress
import io.github.rosemoe.sora.lsp.events.rename.prepareRename
import io.github.rosemoe.sora.lsp.events.rename.rename
import io.github.rosemoe.sora.lsp.events.workspace.setTrace
import io.github.rosemoe.sora.lsp.requests.Timeout
import io.github.rosemoe.sora.lsp.requests.Timeouts
import io.github.rosemoe.sora.lsp.utils.FileUri
import io.github.rosemoe.sora.lsp.utils.clearVersions
import io.github.rosemoe.sora.text.CharPosition
import io.github.rosemoe.sora.widget.CodeEditor
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.future.future
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import org.eclipse.lsp4j.CodeAction
import org.eclipse.lsp4j.ColorInformation
import org.eclipse.lsp4j.Command
import org.eclipse.lsp4j.Diagnostic
import org.eclipse.lsp4j.DocumentHighlight
import org.eclipse.lsp4j.Hover
import org.eclipse.lsp4j.Location
import org.eclipse.lsp4j.LocationLink
import org.eclipse.lsp4j.Position
import org.eclipse.lsp4j.PrepareRenameDefaultBehavior
import org.eclipse.lsp4j.PrepareRenameResult
import org.eclipse.lsp4j.Range
import org.eclipse.lsp4j.SignatureHelp
import org.eclipse.lsp4j.TextDocumentSyncKind
import org.eclipse.lsp4j.WorkspaceEdit
import org.eclipse.lsp4j.jsonrpc.messages.Either
import org.eclipse.lsp4j.jsonrpc.messages.Either3
import java.lang.ref.WeakReference
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeoutException

class LspEditor(
    val project: LspProject,
    val uri: FileUri,
    val languageId: String? = null
) {
    private val delegate = LspEditorDelegate(this)
    internal val uiDelegate = LspEditorUIDelegate(this)

    private var _currentEditor: WeakReference<CodeEditor?> = WeakReference(null)

    private var currentLanguage: LspLanguage? = null

    @Volatile
    private var isClosed = false

    private val disposeLock = Any()

    val eventManager = LspEventManager(project, this)

    val fileExt = uri.path.substringAfterLast('.')

    var textDocumentSyncKind = TextDocumentSyncKind.Incremental

    var completionTriggers = mutableSetOf<String>()

    var signatureHelpTriggers = mutableSetOf<String>()

    var signatureHelpReTriggers = mutableSetOf<String>()

    val coroutineScope = project.coroutineScope

    var editor: CodeEditor?
        set(currentEditor) {
            if (currentEditor == null) {
                throw IllegalArgumentException("Editor cannot be null")
            }

            uiDelegate.detachEditor()
            _currentEditor = WeakReference(currentEditor)

            currentEditor.setEditorLanguage(currentLanguage)
            uiDelegate.attachEditor(currentEditor)
        }
        get() {
            return _currentEditor.get()
        }

    var editorContent: String
        get() = editor?.text?.toString() ?: ""
        set(content) {
            editor?.setText(content)
        }

    var wrapperLanguage: Language? = null
        set(language) {
            field = language
            currentLanguage?.wrapperLanguage = wrapperLanguage
            val editor = _currentEditor.get()
            if (editor != null) {
                this.editor = editor
            }
        }

    var eventListener: LspEditorEventListener = LspEditorEventListener.DEFAULT

    var status: LspEditorStatus = LspEditorStatus.IDLE
        private set(value) {
            if (field == value) return
            val old = field
            field = value
            eventListener.onStatusChanged(this, value, old)
        }

    val isConnected: Boolean
        get() = status == LspEditorStatus.CONNECTED

    val languageServerWrapper: LanguageServerWrapper
        get() = delegate.getPrimaryWrapper()
            ?: throw IllegalStateException("No language server wrapper for language ${languageId ?: fileExt}")

    var diagnostics
        get() = project.diagnosticsContainer.getDiagnostics(uri)
        set(value) {
            publishDiagnostics(value)
        }

    val diagnosticsContainer
        get() = project.diagnosticsContainer

    val isShowSignatureHelp
        get() = uiDelegate.isShowSignatureHelp

    val isShowHover
        get() = uiDelegate.isShowHover

    val isShowCodeActions
        get() = uiDelegate.isShowCodeActions

    var isEnableHover: Boolean
        get() = uiDelegate.isEnableHover
        set(value) {
            uiDelegate.isEnableHover = value
        }

    var isEnableSignatureHelp: Boolean
        get() = uiDelegate.isEnableSignatureHelp
        set(value) {
            uiDelegate.isEnableSignatureHelp = value
        }

    @get:Experimental
    @set:Experimental
    var isEnableInlayHint: Boolean
        get() = uiDelegate.isEnableInlayHint
        set(value) {
            uiDelegate.isEnableInlayHint = value
        }

    val hoverWindow
        get() = uiDelegate.hoverWindow

    val codeActionWindow
        get() = uiDelegate.codeActionWindow

    val signatureHelpWindow
        get() = uiDelegate.signatureHelpWindow

    val requestManager
        get() = delegate.aggregatedRequestManager

    val requestManagers: List<RequestManager>
        get() = delegate.aggregatedRequestManager.activeManagers

    init {
        currentLanguage = LspLanguage(this)
    }


    /**
     * Connect to the language server to provide the capabilities, this will cause threads blocking. Note: An error will be thrown if the language server is not connected after some time.
     *
     * @see io.github.rosemoe.sora.lsp.requests.Timeouts
     *
     * @see io.github.rosemoe.sora.lsp.requests.Timeout
     */

    @Throws(TimeoutException::class)
    suspend fun connect(throwException: Boolean = true): Boolean = withContext(Dispatchers.IO) {
        status = LspEditorStatus.CONNECTING
        eventManager.init()
        runCatching {
            // Delegate handles multiserver coordination and returns merged capabilities.
            val capabilities = delegate.connectAll()
                ?: throw TimeoutException("Unable to connect language server")

            openDocument()

            editor?.let { uiDelegate.attachEditor(it) }

            currentLanguage?.let { language ->
                if (capabilities.documentFormattingProvider?.left != false || capabilities.documentFormattingProvider?.right != null) {
                    language.formatter = LspFormatter(language)
                }
            }

            if (capabilities.inlayHintProvider?.left != false || capabilities.inlayHintProvider?.right != null) {
                requestInlayHint(CharPosition(0, 0))
            }
            requestDocumentColor()

            status = LspEditorStatus.CONNECTED
        }.onFailure {
            if (throwException) {
                status = LspEditorStatus.DISCONNECTED
                throw it
            }
            status = LspEditorStatus.DISCONNECTED
        }.isSuccess
    }

    @WorkerThread
    fun connectBlocking(throwException: Boolean = true): Boolean = runBlocking {
        connect(throwException)
    }

    /**
     * Try to connect to the language server repeatedly, this will cause threads blocking. Note: An error will be thrown if the language server is not connected after some time.
     *
     * @see io.github.rosemoe.sora.lsp.requests.Timeouts
     *
     * @see io.github.rosemoe.sora.lsp.requests.Timeout
     */
    @Throws(InterruptedException::class, TimeoutException::class)
    suspend fun connectWithTimeout() {
        var isConnected = false

        var start = System.currentTimeMillis()
        val retryTime = Timeout[Timeouts.INIT, this]
        val maxRetryTime: Long = start + retryTime

        while (start < maxRetryTime) {
            try {
                connect()
                isConnected = true
                break
            } catch (exception: java.lang.Exception) {
                exception.printStackTrace();
            }
            start = System.currentTimeMillis()
            delay((retryTime / 200).toLong())
        }

        if (!isConnected && start > maxRetryTime) {
            throw TimeoutException("Unable to connect language server")
        } else if (!isConnected) {
            connect()
        }

    }

    @WorkerThread
    fun connectWithTimeoutBlocking() = runBlocking {
        connectWithTimeout()
    }

    /**
     * disconnect to the language server
     */
    @WorkerThread
    @Throws(RuntimeException::class)
    fun disconnect() {
        uiDelegate.detachEditor()
        runCatching {
            coroutineScope.future {
                eventManager.emitAsync(EventType.documentClose)
            }.get()
            delegate.disconnectAll()
            status = LspEditorStatus.DISCONNECTED
        }.onFailure {
            status = LspEditorStatus.DISCONNECTED
            delegate.disconnectAll()
            throw it
        }
    }

    internal fun onWrapperStopped(wrapper: LanguageServerWrapper) {
        uiDelegate.clearWrapperState()
        delegate.onWrapperDisconnected(wrapper)
    }

    /**
     * Notify the language server to open the document
     */
    suspend fun openDocument() {
        eventManager.emitAsync(EventType.documentOpen)
    }

    @WorkerThread
    fun openDocumentBlocking() = runBlocking {
        openDocument()
    }

    /**
     * Notify language servers the document is saved
     */
    suspend fun saveDocument() {
        eventManager.emitAsync(EventType.documentSave)
    }

    @WorkerThread
    fun saveDocumentBlocking() = runBlocking {
        saveDocument()
    }

    /**
     * Request the definition of the symbol at the given position
     */
    suspend fun requestDefinition(position: Position?): Either<List<Location>, List<LocationLink>>? {
        return eventManager.emitAsync(EventType.definition) {
            put(position)
        }.getOrNull("result")
    }

    /**
     * Request the definition of the symbol at the given position
     */
    suspend fun requestDefinition(position: CharPosition): Either<List<Location>, List<LocationLink>>? {
        return eventManager.emitAsync(EventType.definition) {
            put(position)
        }.getOrNull("result")
    }

    /**
     * Request the definition of the symbol at the current cursor position
     */
    suspend fun requestDefinition(): Either<List<Location>, List<LocationLink>>? {
        return eventManager.emitAsync(EventType.definition).getOrNull("result")
    }

    /**
     * Request the references of the symbol at the given position
     */
    suspend fun requestReferences(position: Position?, includeDeclaration: Boolean = true): List<Location?>? {
        return eventManager.emitAsync(EventType.references) {
            put(position)
            put("includeDeclaration", includeDeclaration)
        }.getOrNull("result")
    }

    /**
     * Request the references of the symbol at the given position
     */
    suspend fun requestReferences(position: CharPosition, includeDeclaration: Boolean = true): List<Location?>? {
        return eventManager.emitAsync(EventType.references) {
            put(position)
            put("includeDeclaration", includeDeclaration)
        }.getOrNull("result")
    }

    /**
     * Request the references of the symbol at the current cursor position
     */
    suspend fun requestReferences(includeDeclaration: Boolean = true): List<Location?>? {
        return eventManager.emitAsync(EventType.references) {
            put("includeDeclaration", includeDeclaration)
        }.getOrNull("result")
    }

    /**
     * Request the prepare rename of the symbol at the given position
     */
    suspend fun requestPrepareRename(position: Position?): Either3<Range, PrepareRenameResult, PrepareRenameDefaultBehavior>? {
        return eventManager.emitAsync(EventType.prepareRename) {
            put(position)
        }.getOrNull("result")
    }

    /**
     * Request the prepare rename of the symbol at the given position
     */
    suspend fun requestPrepareRename(position: CharPosition): Either3<Range, PrepareRenameResult, PrepareRenameDefaultBehavior>? {
        return eventManager.emitAsync(EventType.prepareRename) {
            put(position)
        }.getOrNull("result")
    }

    /**
     * Request the prepare rename of the symbol at the current cursor position
     */
    suspend fun requestPrepareRename(): Either3<Range, PrepareRenameResult, PrepareRenameDefaultBehavior>? {
        return eventManager.emitAsync(EventType.prepareRename).getOrNull("result")
    }

    /**
     * Request the rename of the symbol at the given position
     */
    suspend fun requestRename(newName: String, position: Position?): WorkspaceEdit? {
        return eventManager.emitAsync(EventType.rename) {
            put("newName", newName)
            put(position)
        }.getOrNull("result")
    }

    /**
     * Request the rename of the symbol at the given position
     */
    suspend fun requestRename(newName: String, position: CharPosition): WorkspaceEdit? {
        return eventManager.emitAsync(EventType.rename) {
            put("newName", newName)
            put(position)
        }.getOrNull("result")
    }

    /**
     * Request the rename of the symbol at the current cursor position
     */
    suspend fun requestRename(newName: String): WorkspaceEdit? {
        return eventManager.emitAsync(EventType.rename) {
            put("newName", newName)
        }.getOrNull("result")
    }

    /**
     * Cancel a progress operation on the server
     */
    suspend fun cancelProgress(token: Either<String, Int>) {
        eventManager.emitAsync(EventType.cancelProgress) {
            put("token", token)
        }
    }

    /**
     * Set the trace setting of the server
     */
    suspend fun setTrace(value: String) {
        eventManager.emitAsync(EventType.setTrace) {
            put("value", value)
        }
    }

    /**
     * Send a custom notification to the server
     */
    fun notify(method: String, parameter: Any?) {
        requestManager.notify(method, parameter)
    }

    /**
     * Send a custom request to the server
     */
    fun request(method: String, parameter: Any?): CompletableFuture<*> {
        return requestManager.request(method, parameter)
    }

    fun onDiagnosticsUpdate() {
        publishDiagnostics(diagnostics)
    }

    private fun publishDiagnostics(diagnostics: List<Diagnostic>) {
        eventManager.emit(EventType.publishDiagnostics) {
            put("data", diagnostics)
        }
    }

    fun showSignatureHelp(signatureHelp: SignatureHelp?) {
        uiDelegate.showSignatureHelp(signatureHelp)
    }

    fun showHover(hover: Hover?) {
        uiDelegate.showHover(hover)
    }

    fun showCodeActions(range: Range?, actions: List<Either<Command, CodeAction>>?) {
        uiDelegate.showCodeActions(range, actions)
    }

    fun showDocumentHighlight(highlights: List<DocumentHighlight>?) {
        uiDelegate.showDocumentHighlight(highlights)
    }

    internal fun showInlayHints(inlayHints: List<org.eclipse.lsp4j.InlayHint>?) {
        uiDelegate.showInlayHints(inlayHints)
    }

    internal fun showDocumentColors(documentColors: List<ColorInformation>?) {
        uiDelegate.showDocumentColors(documentColors)
    }

    fun getAllColorOccurrences(): List<ColorInformation> {
        return uiDelegate.cachedDocumentColors ?: return emptyList()
    }

    fun hitReTrigger(eventText: CharSequence): Boolean {
        for (trigger in signatureHelpReTriggers) {
            if (trigger.contains(eventText)) {
                return true
            }
        }
        return false
    }

    fun hitTrigger(eventText: CharSequence): Boolean {
        for (trigger in signatureHelpTriggers) {
            if (trigger.contains(eventText)) {
                return true
            }
        }
        return false
    }

    @WorkerThread
    fun dispose() {
        synchronized(disposeLock) {
            if (isClosed) {
                return
                // throw IllegalStateException("Editor is already closed")
            }
            disconnect()
            _currentEditor.clear()
            clearVersions {
                it == this.uri
            }
            project.removeEditor(this)
            isClosed = true
        }
    }

    suspend fun disposeAsync() = withContext(Dispatchers.IO) {
        dispose()
    }
}
